### ISSUE TITLE
Extend the doc comment syntax to allow linking to other gleam code

### ISSUE THREAD

**Issue by george-grec** [gleam-lang/gleam#2869]
Sometimes you want to refer to another specific piece of code when documenting something. You can currently just write the name of the code segment in ticks and the reader can search for that manually but it would be much more useful if there was a standard way to link somewhere and have it be a clickable link in an IDE or on the generated docs page.

Possible inspirations:
- the documentation language for Kotlin [KDoc](https://kotlinlang.org/docs/kotlin-doc.html)
- rustdoc as used in Rust, [Linking to items by name](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html)

Personally, I would love to able to link to functions, types, variants or record accessors.

**Example**
Take the function `fn from_uri(uri: Uri)` from the `gleam_http` package which creates a request from a Uri that you can pass to an http client.
Now there is also the function `fn to(url: String)` that does the exact same thing but you can just give it a String which will be parsed into a Uri and passed on to `from_uri(uri: Uri)`.
A user of this API might wonder what http method the returned request might have. If the documentation for `from_uri` already specifies that it defaults to `GET` then you could write the documentation for `fn to(url: String)` this way:

```gleam
/// Construct a request from a URL string. The returned [request.Request] will have the same defaults
/// as [request.from_uri()] as that is used under the hood.
///
pub fn to(url: String) -> Result(Request(String), Nil)
```

When hovering over `to()` somewhere in your code or when looking at the hexdocs of that function the discoverability would then be much better if you could instantly jump to the referenced function `from_uri()` to read up on it.

**Comment by lpil** [comment]
Good idea! Thank you. We'll need to decide the precise syntax. I think it would be nice if the syntax could be valid markdown too. For example:

```gleam
/// Construct a request from a URL string. The returned `gleam/http/request.Request` will have the same defaults
/// as `gleam/http/request.from_uri` as that is used under the hood.
///
pub fn to(url: String) -> Result(Request(String), Nil)
```

One tricky thing is when a type and a value have the same name. Which should we link to? I guess the syntax would need to be able specify one or the other.

**Comment by george-grec** [comment]
@lpil Good points!
> I think it would be nice if the syntax could be valid markdown too

Strictly speaking using brackets around some text is not invalid markdown but I see where you are coming from. Still, I think it is fine if the code links would just look like regular text with visible brackets in a regular markdown editor. We can expect most gleam code to be open in editors with gleam support where that should not be an issue. Then we could keep backticks as they are for just providing code examples without also being links which I would suggest is a common use case. You could also argue that the functionality that we want (linking to code) is closer to markdown links which also utilize brackets.

> One tricky thing is when a type and a value have the same name. Which should we link to?

I am not sure what you mean by "value" in this context but if you mean something like this:
```gleam
pub type Foo {
  Foo
}
```
then `[request.Foo]` should link to the type and `[request.Foo.Foo]` should link to the variant / constructor.

In any case, I think it would be best to always require the name of the module for every code link. However, if the type is already imported I don't think a full module path (like `gleam/http/request.Request` from your example) is required.
For example:
```gleam
/// see [request.from_uri] // Good
/// see [from_uri] // Bad, even if the function is defined in the same module
///
pub fn to(url: String) -> Result(Request(String), Nil)
```

An exception should be made when referencing an argument in the function that is being documented.
```gleam
/// returns `Nil` if [url] is invalid // Still good, because url is an argument for this function
///
pub fn to(url: String) -> Result(Request(String), Nil)
```
You will notice that I used backticks instead of a link for `Nil` as it is not something that can be linked to ;)

Would this cover all ambiguities?

**Comment by lpil** [comment]
I would want it to ideally be markdown formatting that still makes sense when there's regular markdown rendering or no rendering, so possibly backticks. We can look at what other tools do.

Value constructors are not namespaced by their type so I'm not a huge fan of having them be so in this context. We try and make everything as consistent or explicit as possible.

How would it know that "request" means the "gleam/http/request" module?

**Comment by george-grec** [comment]
> How would it know that "request" means the "gleam/http/request" module?

By being aware of the imports in the current file. Of course, if you reference something from a module that is not imported you would still need to use the full path. That is at least how it works in KDoc.

**Comment by lpil** [comment]
How do we avoid links breaking when imports are changed? To we have to parse the comments and check them on each compile run?

**Comment by george-grec** [comment]
> I would want it to ideally be markdown formatting that still makes sense when there's regular markdown rendering or no rendering, so possibly backticks. We can look at what other tools do.

Sure, I am biased because the tools I use use brackets for this but I'd be happy if other people could make the case for other options so we explore all possibilities. 

---
> Value constructors are not namespaced by their type so I'm not a huge fan of having them be so in this context. We try and make everything as consistent or explicit as possible.

Here is an idea off the top of my head:
```gleam
// in file example.gleam

pub type Foo {
  Foo
}
```

`` /// see `example.{type Foo}` `` references the type and `` /// see `example.Foo` `` references the variant

---
> How do we avoid links breaking when imports are changed? To we have to parse the comments and check them on each compile run?

When writing Kotlin in IntelliJ (and Android Studio I guess) the refactoring tools change references in documentation comments too. If you reference something that is not recognized, the code link in the comment section is highlighted like an error in the editor. To me that sounds like a feature for the LSP and not the compiler but I am not familiar how those things work under the hood - I am just a user of those features.

**Comment by lpil** [comment]
Given the LSP is an optional extra rather than a required tool for writing Gleam any feature needs to work just as well when not using it. We could have the compiler or build tool do something to cover all users, but not the language server.

**Comment by giacomocavalieri** [comment]
Louis recently proposed something like [this](https://discord.com/channels/768594524158427167/768594524158427170/1421129943051534496):
> I think it like like `t:MyType`, `m:some/module`, `v:some_value`, `t:some/module.SomeType`

**Comment by lpil** [comment]
Copied from Elixir!

**Comment by giacomocavalieri** [comment]
Sounds good to me there's still some things that are not totally clear to me on how the compiler should behave:
1. When we perform a rename with the LS I'm guessing it should rename those as well?
2. What should happen if one of the referenced values doesn't exist? Should the compiler raise a warning?
3. Should those values be wrapped in backticks ` to be understood by the compiler? I think it would be nice to ensure that those will look like code even when using a tool that is not aware of this Gleam/Elixir specific convention
4. How should those values be displayed in the generated documentation? Should we use the type printer we're already using that makes sure types and values are always qualified as needed or should we respect the developer's decision on how it has to be displayed?
5. How are these references resolved? When one is not qualified is it expected to be defined in the given module or should we use a module's imports to resolve it?

**Comment by fruno-bulax** [comment]
> When we perform a rename with the LS I'm guessing it should rename those as well?

I think that would be very cool but is not strictly necessary in the first step.

> What should happen if one of the referenced values doesn't exist? Should the compiler raise a warning?

This seems mandatory to me, otherwise people will end up with broken documentation links after every refactor.

> Should those values be wrapped in backticks ` to be understood by the compiler? I think it would be nice to ensure that those will look like code even when using a tool that is not aware of this Gleam/Elixir specific convention

That seems to be what [ExDoc](https://github.com/elixir-lang/ex_doc#elixir-2) is doing and I think is a pretty good idea.

> How should those values be displayed in the generated documentation? Should we use the type printer we're already using that makes sure types and values are always qualified as needed or should we respect the developer's decision on how it has to be displayed?

I'm not really familiar with the type printer, but in general I think it makes more sense if the visible text is pretty-printed. The "fully qualified reference" is embodied by the link already.

> How are these references resolved? When one is not qualified is it expected to be defined in the given module or should we use a module's imports to resolve it?

I'm used to the way [KDoc does it](https://kotlinlang.org/docs/kotlin-doc.html#links-to-elements), where references are aware of imports. That being said, gleam doesn't suffer from the same level of module nesting the java world.

In ExDoc, local elements can be unqualified and everything else must be fully qualified, which I think could also work well for gleam.
Full qualification would also be more understandable when rendered in plain markdown.

**Comment by lpil** [comment]
> When we perform a rename with the LS I'm guessing it should rename those as well?

Good call. I think we'd need to do so? Otherwise renaming would then result in broken code.

> What should happen if one of the referenced values doesn't exist? Should the compiler raise a warning?

We'd want to error and block publishing. For regular development a warning sounds good, though having to parse markdown during analysis isn't great as it would impact compile times and LS responsiveness... Not sure what's the best approach here.

> Should those values be wrapped in backticks ` to be understood by the compiler? I think it would be nice to ensure that those will look like code even when using a tool that is not aware of this Gleam/Elixir specific convention

Sorry, what do you mean?

> How should those values be displayed in the generated documentation? Should we use the type printer we're already using that makes sure types and values are always qualified as needed or should we respect the developer's decision on how it has to be displayed?

I think it should use the global names, not the local ones.

> How are these references resolved? When one is not qualified is it expected to be defined in the given module or should we use a module's imports to resolve it?

Global names. Local code and imports have no impact on documentation comments.

**Comment by giacomocavalieri** [comment]
This could be turned into a discussion since it's not immediately actionable!
